Archives/ClmFile.cpp

```diff
@@ -463,97 +458,94 @@ namespace Archives
 	}
 
 	bool ClmFile::WriteVolume(HANDLE outFile,
-		int numFilesToPack,
 		HANDLE *fileHandle,
 		IndexEntry *entry,
-		std::vector<std::string> internalNames,
+		const std::vector<std::string>& internalNames,
 		WaveFormatEx *waveFormat)
 	{
 #pragma pack(push, 1)
-		struct SClmHeader
+		struct ClmHeader
 		{
-			char textBuff[32];
+			ClmHeader() { memcpy(fileVersion.data(), "OP2 Clump File Version 1.0\x01A\0\0\0\0", 32); }
+
+			std::array<char, 32> fileVersion;
 			WaveFormatEx waveFormat;
-			char unknown[6];
-			int numIndexEntries;
+			std::array<char, 6> unknown { 0, 0, 0, 0, 1, 0 };
+			int packedFilesCount;
 		};
 #pragma pack(pop)
 
-		char buff[CLM_WRITE_SIZE];
-		SClmHeader header;
-		//	char textBuff[32] = "OP2 Clump File Version 1.0\x01A\0\0\0\0";
-		unsigned long numBytesRead;
-		unsigned long numBytes;
-		int i;
-		int offset;
-
-		memcpy(&header.textBuff, "OP2 Clump File Version 1.0\x01A\0\0\0\0", 32);
+		ClmHeader header;
 		header.waveFormat = *waveFormat;
-		memcpy(header.unknown, m_Unknown, 6);
-		header.numIndexEntries = numFilesToPack;
+		header.packedFilesCount = internalNames.size();
 
 		// Write the header
-		if (WriteFile(outFile, &header, sizeof(header), &numBytes, nullptr) == 0) return false;
-
-		/*
-		// Write the text header
-		if (WriteFile(outFile, textBuff, 32, &numBytes, NULL) == 0) return false;
-		// Write the wave format
-		if (WriteFile(outFile, waveFormat, sizeof(WaveFormatEx), &numBytes, NULL) == 0) return false;
-		// Write the unknown bytes
-		if (WriteFile(outFile, m_Unknown, 6, &numBytes, NULL) == 0) return false;
-		// Write the number of internal files
-		if (WriteFile(outFile, &m_NumberOfPackedFiles, 4, &numBytes, NULL) == 0) return false;
-		*/
-
-		// Prepare the index entries
-		offset = sizeof(header) + numFilesToPack * sizeof(IndexEntry);
-		for (i = 0; i < numFilesToPack; i++)
-		{
-			// Copy the filename into the entry
-			strncpy((char*)&entry[i].fileName, internalNames[i].c_str(), 8);
-			// Set the offset of the file
-			entry[i].dataOffset = offset;
-			offset += entry[i].dataLength;
+		unsigned long numBytes;
+		if (WriteFile(outFile, &header, sizeof(header), &numBytes, nullptr) == 0) {
+			return false;
 		}
 
-		// Write the index
-		if (WriteFile(outFile, entry, numFilesToPack * sizeof(IndexEntry), &numBytes, nullptr) == 0)
-		{
+		PrepareIndex(sizeof(header), internalNames, entry);
```

Comment:
Perhaps this method should return a vector of IndexEntry structs, rather than take an array to fill.

Reply:
I've been thinking about this as well. If we plan on only opening one fileStream at a time when writing a ClmFile, we will not be able to set all values of IndexEntry at once. 

We will need to write default values into the ClmFile initially. Then as each file is added, seek back to the header for setting the actual values of each IndexEntry. 

So, I think what I'd like to see is expanding ClmHeader to include the IndexEntry container and look something like:

```
struct ClmHeader
{
    //TODO: Allow setting default values if ClmFile doesn't exist yet.

    // TODO: Check that index is > 0 & less than packedFilesCount
    InsertIndexEntry(int index, IndexEntry entry);

    std::array<char, 32> fileVersion; 
    WaveFormatEx waveFormat; 
    std::array<char, 6> unknown { 0, 0, 0, 0, 1, 0 }; 
    int packedFilesCount; 
    std::vector<IndexEntry> index;
}
```

Reply:
We only need to know the file's name and length for the index entries. Nothing about the file's contents is required. Hence we don't actually need to open any of the files to build the index table.

I'd also like to avoid the whole can of worms of modifying a file in place by inserting new index entries. If the file is tightly packed, data will need to be moved to make room for extra index entries. That's not an efficient operation. Plus, if the program crashes, or there is a power outage during this process, and the modification was done in place, it's likely to leave the file corrupted.

Reply:
I think if we force the user to specify the number of and locations of all files to be packed before beginning to pack, there shouldn't be any need to move data, just fill in default index values that were already reserved in the file. I was planning on determining the size of each file after the stream was opened. If there is a way to determine this without opening a stream against the file, it could simplify the process. 

We would just want to check to ensure the file still exists and is the expected size when the file is actually opened for packing.

However this goes, it will all be done on a different branch than this one though.

Reply:
There are two ways I know of to [determine file size](https://stackoverflow.com/questions/5840148/how-can-i-get-a-files-size-in-c) without having to open the file. Either using <sys/stat> or <experimental/filesystem>. The filesystem method looks like the cleanest, by far, so I'll be happy when the experimental part is dead and buried.

Reply:
For an extra example of using stat, see [How to get filesize using stat](https://techoverflow.net/2013/08/21/how-to-get-filesize-using-stat-in-cc/).